docs/spec/lightclient/detection.md

```diff
@@ -0,0 +1,491 @@
+     initialization. Initially this set is partitioned into
+    -  one full node that is the *primary* (singleton set),
+	-  a set *Secondaries* (of fixed size, e.g., 3),
+	-  a set *FullNodes*.
+- A set *FaultyNodes* of nodes that the light client suspects of being faulty; it is initially empty
+
+- *Lightstore* as defined in the [verification specification][verification].
+
+
+
+
+#### **[LCD-INV-NODES]:**
+The detector shall maintain the following invariants:
+   - *FullNodes \intersect Secondaries = {}*
+   - *FullNodes \intersect FaultyNodes = {}*
+   - *Secondaries \intersect FaultyNodes = {}*
+   
+and the following transition invariant
+   - *FullNodes' \union Secondaries' \union FaultyNodes' = FullNodes \union Secondaries \union FaultyNodes*
+
+## Solution
+
+
+
+### Inter Process Communication
+
+
+```go
+func FetchLightBlock(peer PeerID, height Height) LightBlock
+```
+See the [verification specification][verification] for details.
+
+### Auxiliary Functions (Local)
+
+
+```go
+Replace_Secondary(addr Address)
+```
+- Expected precondition
+    - *FullNodes* is nonempty
+- Expected postcondition
+    - addr is moved from *Secondaries* to *FaultyNodes*
+    - an address *a* is moved from *FullNodes* to *Secondaries*
+- Error condition
+    - if precondition is violated
+
+
+
+### From the verifier
+```go
+func VerifyToTarget(primary PeerID, lightStore LightStore,
+                    targetHeight Height) (LightStore, Result)
+```
+See the [verification specification][verification] for details.
+
+
+
+## Solution
+
+Shared data of the light client
+- a pool of full nodes *FullNodes* that have not been contacted before
+- peer set called *Secondaries*
+- primary
+- LightStore
+
+
+
+
+The problem is solved by calling  the function `ForkDetector` with
+a lightstore that contains a light block that has
+just been verified by the verifier. 
+
+
+```go
+func ForkDetector(ls LightStore)  {
+    Forks.Init // initialize a container in which we collect forks
+	for i, s range Secondaries {
+		sh := FetchLightBlock(s,LightStore.LatestVerified().Height)
+		// as the check below only needs the header, it is sufficient
+		// to download the header rather than the LighBlock
+		if LightStore.LatestVerified().Header == sh.Header {
+				// header matches. we do nothing.
+	    }
+		else {
+			    // [LCD-REQ-REP]
+			    // header does not match. there is a situation.
+				// we try to verify sh by querying s
+				// we set up an auxiliary lightstore with the highest
+			    // trusted lightblock and the lightblock we want to verify
+				auxLS.Init
+				auxLS.Update(LightStore.LatestTrusted(), StateVerified);
+				auxLS.Update(sh,StateUnverified);
+				result := VerifyToTarget(s, auxLS, LightStore.LatestVerified().Height)
+				if result = (_,ResultSuccess) || (_,EXPIRED) {
+				    // we verified header sh which is conflicting to hd
+					// there is a fork on the main blockchain.
+                    // If return code was EXPIRED it might be too late
```

Comment:
This shouldn't be the case. We should ensure that even if it timeout we can still have time to punish it.

Reply:
I am not sure we can enforce this. If the detection starts late, and is based on old light blocks, this might always happen. I have added a TODO to check whether we can do something there.

Reply:
That's good point. I was only thinking about going forward case. But if you ask for older height (for example from RPC endpoint) that could be problematic. These timing assumptions are really tricky. I am more convinced that we need some kind of simplistic approach where we don't provide any "service" for headers that are not within some constant of end of expiration period. It basically means enforcing synchrony assumptions at the request time, and being sure that we are within big safe period (order of days) so we never play close to the expiration point. And we do this at the trust period and then diff between trust and unbonding period gives time for evidence processing and fork accountability.